datafusion/ffi/src/execution_plan.rs

```diff
@@ -0,0 +1,373 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::{ffi::c_void, pin::Pin, sync::Arc};
+
+use abi_stable::{
+    std_types::{RResult, RStr, RString, RVec},
+    StableAbi,
+};
+use arrow::ffi_stream::FFI_ArrowArrayStream;
+use datafusion::error::Result;
+use datafusion::{
+    error::DataFusionError,
+    execution::{SendableRecordBatchStream, TaskContext},
+    physical_plan::{DisplayAs, ExecutionPlan, PlanProperties},
+};
+
+use crate::{
+    plan_properties::{FFI_PlanProperties, ForeignPlanProperties},
+    record_batch_stream::FFI_RecordBatchStream,
+};
+
+#[repr(C)]
+#[derive(Debug, StableAbi)]
+pub struct WrappedArrayStream(#[sabi(unsafe_opaque_field)] pub FFI_ArrowArrayStream);
+
+#[repr(C)]
+#[derive(Debug, StableAbi)]
+#[allow(missing_docs)]
+#[allow(non_camel_case_types)]
+pub struct FFI_ExecutionPlan {
```

Comment:
I really like the fact that this change follows the same setup as arrow-rs's FFI. It makes it familiar to use.